plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go

```diff
@@ -0,0 +1,248 @@
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+)
+
+// New returns a NodeInfoCache implementation.
+// It automatically starts a go routine that manages expiration of assumed pods.
+// "ttl" is how long the assumed pod will get expired.
+// "period" is how long the background goroutine should wait before cleaning up expired pods periodically.
+// "stop" is the channel that signals stopping and we would close background goroutines.
+func New(ttl, period time.Duration, stop chan struct{}) schedulercache.NodeInfoCache {
+	cache := newNodeInfoCache(ttl, period, stop)
+	cache.run()
+	return cache
+}
+
+// This is copied from client/cache.MetaNamespaceKeyFunc .
+// Since pod is ensured to have object, we don't need to take care of the error returned.
+// A more appropriate way is to support Key() method in pod object.
+func podKeyFunc(pod *api.Pod) string {
+	meta := pod.GetObjectMeta()
+	if len(meta.GetNamespace()) > 0 {
+		return meta.GetNamespace() + "/" + meta.GetName()
+	}
+	return meta.GetName()
+}
+
+type nodeInfoCache struct {
+	stop chan struct{}
+
+	mu          sync.Mutex
+	ttl         time.Duration
+	period      time.Duration
+	assumedPods map[string]assumedPod
+	// A map of added pods
+	// - support RemovePodByKey
+	// - update: subtracts information from old pod
+	// - record pod state and implement state machine flow
+	pods      map[string]*api.Pod
+	podStates map[string]podState
+	// TODO: we should also watch node deletion and clean up node entries
+	nodes map[string]*schedulercache.NodeInfo
+}
+
+type podState int
+
+const (
+	podBinded podState = iota + 1
+	podAdded
+	podExpired
+)
+
+type assumedPod struct {
+	pod            *api.Pod
+	expectToExpire time.Time
+}
+
+func newNodeInfoCache(ttl, period time.Duration, stop chan struct{}) *nodeInfoCache {
+	return &nodeInfoCache{
+		ttl:    ttl,
+		period: period,
+		stop:   stop,
+
+		nodes:       make(map[string]*schedulercache.NodeInfo),
+		pods:        make(map[string]*api.Pod),
+		podStates:   make(map[string]podState),
+		assumedPods: make(map[string]assumedPod),
+	}
+}
+
+func (cache *nodeInfoCache) run() {
+	go util.Until(cache.cleanupExpiredAssumedPods, cache.period, cache.stop)
+}
+
+// return nil if no such node is found for given node name.
+func (cache *nodeInfoCache) GetNodeInfo(nodeName string) *schedulercache.NodeInfo {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	n, ok := cache.nodes[nodeName]
+	if !ok {
+		return nil
+	}
+	if n.PodNum == 0 {
+		return nil
+	}
+	return n.Clone()
+}
+
+func (cache *nodeInfoCache) AssumePod(pod *api.Pod) error {
+	return cache.assumePod(pod, time.Now())
+}
+
+// separate out form time deterministic testsability
+func (cache *nodeInfoCache) assumePod(pod *api.Pod, now time.Time) error {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	key := podKeyFunc(pod)
+	if _, ok := cache.podStates[key]; ok {
+		return fmt.Errorf("Pod state wasn't initial but get assumed. Pod key: %v", key)
```

Comment:
pod state was't initial but assumed. Pod key... (error message should not be capitalized)

see https://github.com/golang/go/wiki/CodeReviewComments#error-strings